### ISSUE TITLE
support bulk copy for mssql and synapse

### ISSUE THREAD

**Issue by rudolfix** [dlt-hub/dlt#1234]
**Background**
The way we insert rows into mssql is ineffective. We should switch to bulk copy. MS odbc drivers come with `bcp` command that we can use.
https://github.com/yehoshuadimarsky/bcpandas/blob/master/bcpandas/utils.py
is a project that does that quite well

**Tasks**
1. * [ ] allow `mssql` and `synapse` to handle `csv`
2. * [ ] use `bcp` in copy jobs

**Comment by jorritsandbrink** [comment]
@rudolfix Do we also want to do this for `synapse`? For `synapse` we already implement `SynapseCopyFileLoadJob` based on `COPY INTO`, which is the recommended approach:

_"While dedicated SQL pools support many loading methods, including popular SQL Server options such as [bcp](https://learn.microsoft.com/en-us/sql/tools/bcp-utility?toc=/azure/synapse-analytics/sql-data-warehouse/toc.json&view=azure-sqldw-latest&preserve-view=true) and the [SqlBulkCopy API](https://learn.microsoft.com/en-us/dotnet/api/system.data.sqlclient.sqlbulkcopy?toc=/azure/synapse-analytics/sql-data-warehouse/toc.json), the fastest and most scalable way to load data is through PolyBase external tables and the [COPY statement](https://learn.microsoft.com/en-us/sql/t-sql/statements/copy-into-transact-sql?toc=/azure/synapse-analytics/sql-data-warehouse/toc.json&view=azure-sqldw-latest&preserve-view=true)._

_With PolyBase and the COPY statement, you can access external data stored in Azure Blob storage or Azure Data Lake Store via the T-SQL language. For the most flexibility when loading, we recommend using the COPY statement._"

Source: https://learn.microsoft.com/en-us/azure/synapse-analytics/sql-data-warehouse/design-elt-data-loading

**Comment by rudolfix** [comment]
@jorritsandbrink it just may work for synapse if we implement it ie. via csv files + bcp but it is not required. The reason for the ticket is abysmal insert performance of sql server. it is WAY slower than postgres. nevertheless priority of this ticket is quit low (at least for now)

**Comment by cody-scott** [comment]
Chiming in here; one of the big hold back for shifting to dlt is the abysmal odbc performance over bcp. 

Some of our sources are tall and wide so bcp helps a ton with performance, so if this comes to be then I would be thrilled!

More considerations about the nuances of the data though (date time/floats/column ordering/etc) then sqlalchemy.

**Comment by bhuesemann** [comment]
Hi @rudolfix , 
for a mssql destination: apart from loading in parallel/batches and current approach of arrayed inserts or using bcp for bulk loading, would it possible to also use single prepared/parameterized inserts and using fast_executemany driver options? This would leave handling of arrayed inserts to the driver and may lead to better bulk database loading performance.
Something like this:
```
cnxn = pyodbc.connect(connection_string, autocommit=True)
crsr = cnxn.cursor()
data = crsr.execute("SELECT * FROM MillionRows").fetchall()
crsr.close()
printmem("data loaded")
crsr = cnxn.cursor()
ins = "INSERT INTO #tmp (ID, TextField, LongIntegerField, DoubleField, varchar_column) VALUES (?,?,?,?,?)"
crsr.fast_executemany = True
crsr.executemany(ins, data)
crsr.close()
```
I could create an sqlalchemy engine like this:
engine = create_engine(destination_conn, fast_executemany=True)
But how to get this configuration into a dlt destination? I can't find a way to inject it with a created destination:
destination = dlt.destinations.mssql(credentials)
destination.spec seems not the right way to pass sa engine parameters.

The post here is somewhat old and the current pyodbc version doesn't need to intercept events:
https://medium.com/analytics-vidhya/speed-up-bulk-inserts-to-sql-db-using-pandas-and-python-61707ae41990

But you can see the possible performance gain as compared to other approaches. And meanwhile I also understand, that using prepared statements alone is no improvement, as the network round trips stay the same. So fast_executemany will also batch inserts but I guess will use a real arrayed insert, may be even bypassing transaction management (thus autocommit needs to be disabled) like real bulk loading would.

**Comment by burnbay** [comment]
Hi @rudolfix 
Would really like to see @bhuesemann's suggestion investigated.

**Comment by cody-scott** [comment]
I'll add here that there is still going to be a performance limitation for large datasets with odbc vs the bulk copy tool. 

Would be interesting to see the performance comparisons for a larger insert (1 million rows) to compare.

**Comment by AndrewLockton** [comment]
> Chiming in here; one of the big hold back for shifting to dlt is the abysmal odbc performance over bcp.
> 
> Some of our sources are tall and wide so bcp helps a ton with performance, so if this comes to be then I would be thrilled!
> 
> More considerations about the nuances of the data though (date time/floats/column ordering/etc) then sqlalchemy.

Would love to BCP. Needing to move very large oracle tables in MSSQL and having BCP for the import would be magical

**Comment by deschman** [comment]
Wow this is slow! Came here looking for answers about why my job loading 150k CRM contacts got stuck and found this instead. Five minutes for something that should be under 10 seconds with bcp might keep us from implementing for reverse ETL. Would love to see this implemented if anyone has bandwidth.

**Comment by rudolfix** [comment]
@lvanoverberghe for now you could try to use `sqlalchemy` destination with mssql. AFAIK internally it does `fast_executemany` (same thing like pandas).

we can support `fast_excutemany` easily because scaffolding is already there. here are the requirements:
1. support `parquet` and `json-typed` file loader extensions
2. use `DestinationLoadJob` - it should work right away
3. read https://medium.com/@daniar.achakeyev/how-to-load-data-faster-using-pyodbc-into-a-ms-sql-server-database-6c0f538dc9f9 very carefully
4. you'll need to set buffer sizes for VARCHAR and VARBINARY. use `precision` we have in the column schema or if not present: add new config setting to mssql configuration. set default value to max column index size for MSSQL (it is somewhere in the code)
5. switch preferred loader file type to `jsonl-typed` - this will make all MSSQL test to use it
6. tests: test buffer sizes. compare insert speed

NEXT STEP: BULK INSERT in a separate PR. for that we need to use staging with csv files and more or less follow `bcp` requirements (it  is using it underneath)

**Comment by VioletM** [comment]
One feedback I've got is that using `sqlalchemy` destination with **SQL INSERT File Format** for mssql is speeding up the loading 4-5 times

**Comment by adrianbr** [comment]
This came up in a recent community discussion on sling vs dlt

https://www.reddit.com/r/dataengineering/comments/1mi9w04/comment/n731l3z/?context=3

mssql destination is a mid single digit % of our userbase

**Comment by match-gabeflores** [comment]
> mssql destination is a mid single digit % of our userbase

Perhaps it's not being used in dlt _because_ of the poor performance? 

https://github.com/dlt-hub/dlt/issues/1234#issuecomment-2162041333

**Comment by lgwacker** [comment]
https://github.com/dlt-hub/dlt/blob/273420b2574a518a7488443253ab1e0971b136e8/dlt/destinations/impl/mssql/sql_client.py#L156

As a quick win, can't we just set curr.fast_executemany=true ?

**Comment by mehmax** [comment]
> One feedback I've got is that using `sqlalchemy` destination with **SQL INSERT File Format** for mssql is speeding up the loading 4-5 times

I am very interested in your solution for this. Could you give some details of the implementation?

**Comment by lvanoverberghe** [comment]
Why does no one want to work on this? There is seemingly a big ask from the community (maybe top issue) and it is being put on the backlog more and more.

**Comment by rudolfix** [comment]
https://columnar.tech/ implemented adbc driver for mssql and we have a working implementation. you'll be able to directly insert parquet into mssql. this will hopefully speed inserts significantly. we'll post more details here soon

**Comment by rudolfix** [comment]
release - closing

**Comment by match-gabeflores** [comment]
> https://columnar.tech/ implemented adbc driver for mssql and we have a working implementation. you'll be able to directly insert parquet into mssql. this will hopefully speed inserts significantly. we'll post more details here soon

do we know how this compares with BULK COPY/INSERT?
